waterbutler/providers/onedrive/provider.py

```
@@ -0,0 +1,419 @@
+import json
+import http
+import asyncio
+
+import logging
+
+from urllib.parse import urlparse
+
+from itertools import repeat
+
+from waterbutler.core import streams
+from waterbutler.core import provider
+from waterbutler.core import exceptions
+#  from waterbutler.tasks.core import backgroundify
+from waterbutler.core.path import WaterButlerPath
+
+from waterbutler.providers.onedrive import settings
+from waterbutler.providers.onedrive.metadata import OneDriveRevision
+from waterbutler.providers.onedrive.metadata import OneDriveFileMetadata
+from waterbutler.providers.onedrive.metadata import OneDriveFolderMetadata
+
+logger = logging.getLogger(__name__)
+
+
+class OneDriveProvider(provider.BaseProvider):
+    """Provider for the OneDrive cloud storage service.
+
+    API docs: https://dev.onedrive.com/README.htm
+    """
+    NAME = 'onedrive'
+    BASE_URL = settings.BASE_URL
+
+    def __init__(self, auth, credentials, settings):
+        super().__init__(auth, credentials, settings)
+        self.token = self.credentials['token']
+        self.folder = self.settings['folder']
+        logger.debug("__init__ credentials:{} settings:{}".format(repr(credentials), repr(settings)))
+
+    async def validate_v1_path(self, path, **kwargs):
+        if path == '/':
+            return WaterButlerPath(path, prepend=self.folder)
+
+        logger.info('validate_v1_path self::{} path::{}  url:{}'.format(repr(self), repr(path), self.build_url(path)))
+
+        resp = await self.make_request(
+            'GET', self.build_url(path),
+            expects=(200, 400),
+            throws=exceptions.MetadataError
+        )
```

What does the 400 response code indicate?  Also, `validate_v1_path` should throw a 404 error if `path` doesn't exist.